jms/1.8/modules/ROOT/pages/jms-troubleshooting.adoc

```diff
@@ -78,6 +78,130 @@ To resolve this scenario:
+* Thread A: consumer A acknowledged message 3 and messages 4,5,6,7,8,9,10,12,14 were returned to the broker.
+* Thread B: consumer B acknowledged message 15 and messages 16 to 24 were returned to the broker.
+* Messages 3 and 15 were consumed.
+
+Workaround: +
+Set the *Max Concurrency* field value of the flow to `1` to force the executions to be serialized and the order of the consumed messages will be as expected.
+
+=== Multiple Consumes on the Same Flow
+
+Conditions +
+
+* Messages are consumed in a foreach block from 1 to 4.
+* Messages consumed in the first point are added to a collection of messages that are later processed and acknowledged.
+* The JMS driver/broker prefetch size is configured to 2.
+* There are 4 messages on the queue.
+* The JMS *Consume* operation *Ack mode* field is configured to `MANUAL` and messages are acknowledged in a foreach block using the collection created in the first point.
+
+Flow iteration 1:
+
+* Foreach iteration1 1:
+** Consumer 1 is created and in its local buffer there are 2 messages ready to process pushed by the broker (messages 1 and 2).
+** Message 1 is added to the collection.
+
+* Foreach iteration 2:
+** Consumer 2 is created and in its local buffer there are 2 messages ready to process pushed by the broker (messages 3 and 4).
+** Message 3 is added to the collection.
+
+* Foreach iteration 3:
+** No available messages in the queue.
+
+* Foreach iteration 4:
+** No available messages in the queue.
+
+Messages on the collection created by the foreach are acknowledged (1,3) and the rest of the messages are returned to the queue (2,4).
+
+Flow iteration 2:
+
+* Foreach iteration 1:
+** Consumer 1 is created and in its local buffer there are 2 messages ready to process pushed by the broker (messages 2 and 4).
+** Message 2 is added to the collection.
+
+* Foreach iteration 2:
+** No available messages in the queue.
+
+* Foreach iteration 3:
+** No available messages in the queue.
+
+* Foreach iteration 4:
+** No available messages in the queue.
+
+Messages on the collection created by the foreach are acknowledged (2) and the rest of the messages are returned to the queue (4).
+
+Flow iteration 3:
+
+* Foreach iteration 1:
+** Consumer 1 is created and in its local buffer there is 1 message ready to process pushed by the broker (message 4).
+** Message 4 is added to the collection.
+
+* Foreach iteration 2:
+** No available messages in the queue.
+
+* Foreach iteration 3:
+** No available messages in the queue.
+
+* Foreach iteration 4:
+** No available messages in the queue.
+
+Messages on the collection created by the foreach are acknowledged (4)
+
+Summarizing the flow iterations, the 3 iterations of the flow consume 4 messages in the queue and the order is not honored:
+
+* Iteration 1: [1,3]
+
+* Iteration 2: [2]
+
+* Iteration 3: [4]
+
+Workaround:
+
+* Disable prefetch on the broker, check your third-party JMS vendor provider documentation to disable prefetch.
```

Comment:
Maybe rewrite like this:

Disable prefetch on the broker, and check your third-party JMS vendor provider documentation to disable prefetch.